### DISCUSSION TITLE
Marking packages as `dev` packages to assume `--dev`

### DISCUSSION THREAD

**Discussion by doekenorg** [composer/composer#10939] [General]
Hi, I had a simple idea. Couldn't find it in the discussions or issues; so if figured I'd post it here. If this is the wrong place I'm terribly sorry.

Some packages are explicitly meant to be `dev` dependencies (testing frameworks, static analyzers, tools that change accessibility of private variables, etc). There is however no system in place to mark these packages as such (of which I am aware). 

I thought it could be a nice feature if a package can mark itself as a `dev` dependency by adding a `dev-` prefix to the `type` parameter (eg. `dev-library`); or introducing a new `dev-package: true` for example.

When requiring such a package with `composer require acme/dev-package` it would assume `--dev`. And when overwriting with `--no-dev` it would install it as a regular dependency; but still issue a warning about the package type.

I think this can help avoid packages from ending up in the wrong list; creating bigger vendor folders than necessary, or actually using these packages, in production.

**Comment by Seldaek** [comment]
I'm not sure about this. Like I totally see it for the 10-15 common testing/ci/static analysis tools. But then your example about accessing private vars im sure many use this in prod and while the author may not have intended it that's still a fact. Now people may require it and have things go in require-dev unexpectedly suddenly.. They may not even notice until its too late and prod breaks.

So it sounds like a bunch of effort for a small gain and possible added confusion too. To me that is overall not worth it.

**Reply by stof** [reply]
@Seldaek things would not suddenly break if this affects only the behavior of `composer require`

**Reply by Seldaek** [reply]
Well the require behavior would suddenly change, potentially breaking things for people down the line.

**Reply by doekenorg** [reply]
It could be introduced like the composer plugins. With a question, if they actually ment `--dev`. And ignore with `-n`.

But the point is that people would opt-in to an unintended usage, rather than by accident.

**Comment by GromNaN** [comment]
It would be better to **suggest** installing a package as dev-dependency when running composer require without the --dev option. That way it's opt-in and in context.

That's something that appens to me quite often. It run `composer req <some dev package>` and then have to edit the composer.json because I forgot the --dev flag. If there was an interactive question like "The author flagged the package for dev, install it as a Dev dependency? y/N

**Reply by Seldaek** [reply]
Yeah that I could live with.. but it's still a bunch of work to propagate a new flag throughout packagist.org etc, and I wonder if the "benefit" is worth it.

**Reply by GromNaN** [reply]
That could leverage an existing config like package type or keywords

**Reply by Seldaek** [reply]
Package type is the wrong fit IMO but a keyword perhaps.. `testing` seems to be quite broadly used on test packages, which could be included in this recommendation. And then we could recommend if `dev` is added too? Or something more specific?

**Reply by Seldaek** [reply]
See https://github.com/composer/composer/pull/10960

**Reply by GuySartorelli** [reply]
I would recommend against keywords for this purpose. Akaik keywords are used only for finding packages, right? I would be surprised if adding a keyword changed the way composer interacts with my package when using `require` - and people may use whatever keyword triggers this without intending it to be a Dev dependency. 

Edit: nvm I see you've already implemented that. Oh well.

**Comment by Ayesh** [comment]
The idea of Composer automatically assuming I wanted to imply `--dev` when I `composer require phpunit/phpunit` certainly has crossed my mind too, but like Jordi said, I imagine it to be a lot work with comparably small impact to implement this feature. 

I think the easiest way would be rolling a small Composer plugin to either lookup a global `composer.json` configuration entry, or use a list of package tags to make this assumption.

Taking and acting on this information from the requested package itself or the repository (as opposed to a configuration initiated by the package consumer) sounds like a lot more work and possibly in footgunning territory most of these tools are used in CI environments, and interactive prompts, notices, etc can disrupt CI builds.

**Reply by Seldaek** [reply]
This is anyway just a prompt happening on interactive sessions, so CI will not be affected at all. See https://github.com/composer/composer/pull/10960
